From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Dmitry Muhomor <muhomor.dmitry@gmail.com>
Date: Wed, 1 Nov 2023 21:27:18 +0200
Subject: [PATCH] never enable GWP-ASan

It weakens hardened_malloc protections, especially when memory tagging is enabled.
---
 libc/bionic/android_mallopt.cpp       |  8 --------
 libc/bionic/gwp_asan_wrappers.h       |  7 -------
 libc/bionic/malloc_common_dynamic.cpp |  2 --
 tests/malloc_test.cpp                 | 23 -----------------------
 4 files changed, 40 deletions(-)

diff --git a/libc/bionic/android_mallopt.cpp b/libc/bionic/android_mallopt.cpp
index 79e407237..d6d0257b1 100644
--- a/libc/bionic/android_mallopt.cpp
+++ b/libc/bionic/android_mallopt.cpp
@@ -66,14 +66,6 @@ extern "C" bool android_mallopt(int opcode, void* arg, size_t arg_size) {
     *reinterpret_cast<bool*>(arg) = atomic_load(&__libc_globals->decay_time_enabled);
     return true;
   }
-  if (opcode == M_INITIALIZE_GWP_ASAN) {
-    if (arg == nullptr || arg_size != sizeof(android_mallopt_gwp_asan_options_t)) {
-      errno = EINVAL;
-      return false;
-    }
-
-    return EnableGwpAsan(*reinterpret_cast<android_mallopt_gwp_asan_options_t*>(arg));
-  }
   if (opcode == M_MEMTAG_STACK_IS_ON) {
     if (arg == nullptr || arg_size != sizeof(bool)) {
       errno = EINVAL;
diff --git a/libc/bionic/gwp_asan_wrappers.h b/libc/bionic/gwp_asan_wrappers.h
index 219da9fc5..9053d4aa4 100644
--- a/libc/bionic/gwp_asan_wrappers.h
+++ b/libc/bionic/gwp_asan_wrappers.h
@@ -35,13 +35,6 @@
 #include "private/bionic_globals.h"
 #include "private/bionic_malloc_dispatch.h"
 
-// Enable GWP-ASan, used by android_mallopt. Should always be called in a
-// single-threaded context.
-bool EnableGwpAsan(const android_mallopt_gwp_asan_options_t& options);
-
-// Hooks for libc to possibly install GWP-ASan.
-bool MaybeInitGwpAsanFromLibc(libc_globals* globals);
-
 // Returns whether GWP-ASan is the provided dispatch table pointer. Used in
 // heapprofd's signal-initialization sequence to determine the intermediate
 // dispatch pointer to use when initing.
diff --git a/libc/bionic/malloc_common_dynamic.cpp b/libc/bionic/malloc_common_dynamic.cpp
index 330415be8..12f9c85cf 100644
--- a/libc/bionic/malloc_common_dynamic.cpp
+++ b/libc/bionic/malloc_common_dynamic.cpp
@@ -383,8 +383,6 @@ static void MallocInitImpl(libc_globals* globals) {
   char prop[PROP_VALUE_MAX];
   char* options = prop;
 
-  MaybeInitGwpAsanFromLibc(globals);
-
 #if defined(USE_SCUDO) && !__has_feature(hwaddress_sanitizer)
   __libc_shared_globals()->scudo_stack_depot = __scudo_get_stack_depot_addr();
   __libc_shared_globals()->scudo_region_info = __scudo_get_region_info_addr();
diff --git a/tests/malloc_test.cpp b/tests/malloc_test.cpp
index 3f1ba7959..767fd439d 100644
--- a/tests/malloc_test.cpp
+++ b/tests/malloc_test.cpp
@@ -1412,29 +1412,6 @@ TEST(android_mallopt, set_allocation_limit_multiple_threads) {
 #endif
 }
 
-#if defined(__BIONIC__)
-using Mode = android_mallopt_gwp_asan_options_t::Mode;
-TEST(android_mallopt, DISABLED_multiple_enable_gwp_asan) {
-  android_mallopt_gwp_asan_options_t options;
-  options.program_name = "";  // Don't infer GWP-ASan options from sysprops.
-  options.mode = Mode::APP_MANIFEST_NEVER;
-  // GWP-ASan should already be enabled. Trying to enable or disable it should
-  // always pass.
-  ASSERT_TRUE(android_mallopt(M_INITIALIZE_GWP_ASAN, &options, sizeof(options)));
-  options.mode = Mode::APP_MANIFEST_DEFAULT;
-  ASSERT_TRUE(android_mallopt(M_INITIALIZE_GWP_ASAN, &options, sizeof(options)));
-}
-#endif  // defined(__BIONIC__)
-
-TEST(android_mallopt, multiple_enable_gwp_asan) {
-#if defined(__BIONIC__)
-  // Always enable GWP-Asan, with default options.
-  RunGwpAsanTest("*.DISABLED_multiple_enable_gwp_asan");
-#else
-  GTEST_SKIP() << "bionic extension";
-#endif
-}
-
 TEST(android_mallopt, memtag_stack_is_on) {
 #if defined(__BIONIC__)
   bool memtag_stack;
